google/cloud/spanner_v1/merge_result_set.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2024 Google LLC All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from queue import Queue
+from typing import Any, TYPE_CHECKING
+from threading import Lock, Event
+
+if TYPE_CHECKING:
+    from google.cloud.spanner_v1.database import BatchSnapshot
+
+QUEUE_SIZE_PER_WORKER = 32
+MAX_PARALLELISM = 16
+METADATA_LOCK = Lock()
+
+
+def _set_metadata(merged_result_set, results):
+    METADATA_LOCK.acquire()
+    try:
+        merged_result_set._metadata = results.metadata
+    finally:
+        METADATA_LOCK.release()
+        merged_result_set.metadata_event.set()
+
+
+class PartitionExecutor:
+    """
+    Executor that executes single partition on a separate thread and inserts
+    rows in the queue
+    """
+
+    def __init__(self, batch_snapshot, partition_id, merged_result_set):
+        self._batch_snapshot: BatchSnapshot = batch_snapshot
+        self._partition_id = partition_id
+        self._merged_result_set: MergedResultSet = merged_result_set
+        self._queue: Queue[PartitionExecutorResult] = merged_result_set._queue
+
+    def run(self):
+        try:
+            results = self._batch_snapshot.process_query_batch(self._partition_id)
+            merged_result_set = self._merged_result_set
+            for row in results:
+                if merged_result_set._metadata is None:
+                    _set_metadata(merged_result_set, results)
+                self._queue.put(PartitionExecutorResult(data=row))
+            # Special case: The result set did not return any rows.
+            # Push the metadata to the merged result set.
+            if merged_result_set._metadata is None:
+                _set_metadata(merged_result_set, results)
+        except Exception as ex:
+            self._queue.put(PartitionExecutorResult(exception=ex))
```

Comment:
We should also call `_set_metadata` here (if it has not already been set) to prevent the `metadata` property from blocking indefinitely if someone tries to call that after an error has occurred. If for example the query fails for all partitions, then the user will get a `MergedResultSet` that returns an error whenever you try to iterate over the rows, but that hangs forever if you try to call `metadata`. The latter is always very hard to debug, so we should whenever possible return an error instead of block when something goes wrong.

Reply:
Done